includes/Classifai/Admin/BulkActions.php

```diff
@@ -76,29 +110,84 @@ public function bulk_action_handler( $redirect_to, $doaction, $post_ids ) {
 		foreach ( $post_ids as $post_id ) {
 			$this->save_post_handler->classify( $post_id );
 		}
+		$redirect_to = remove_query_arg( [ 'bulk_classified', 'bulk_scanned', 'bulk_cropped' ], $redirect_to );
 		$redirect_to = add_query_arg( 'bulk_classified', count( $post_ids ), $redirect_to );
 		return $redirect_to;
 	}
 
 	/**
-	 * Display an admin notice after classifying posts in bulk.
+	 * Handle media bulk actions.
+	 *
+	 * @param string $redirect_to       Redirect URL after bulk actions.
+	 * @param string $doaction          Action ID.
+	 * @param array  $attachment_ids    Attachment ids to apply bulk actions to.
+	 *
+	 * @return string
+	 */
+	public function media_bulk_action_handler( $redirect_to, $doaction, $attachment_ids ) {
+		if (
+			empty( $attachment_ids ) ||
+			! in_array( $doaction, [ 'scan_image', 'smart_crop' ], true )
+		) {
+			return $redirect_to;
+		}
+
+		foreach ( $attachment_ids as $attachment_id ) {
+			$current_meta = wp_get_attachment_metadata( $attachment_id );
+
+			if ( 'smart_crop' === $doaction ) {
+				$this->computer_vision->smart_crop_image( $current_meta, $attachment_id );
+			} else {
+				$this->computer_vision->generate_image_alt_tags( $current_meta, $attachment_id );
+			}
+		}
+
+		$action = 'scan_image' === $doaction ? 'scanned' : 'cropped';
+
+		$redirect_to = remove_query_arg( [ 'bulk_classified', 'bulk_scanned', 'bulk_cropped' ], $redirect_to );
+		$redirect_to = add_query_arg( "bulk_{$action}", count( $attachment_ids ), $redirect_to );
+		return $redirect_to;
```

Comment:
The dynamic `$action` is safe now but it would be good to add some defensive coding in case a future dev modifies it between lines 145 & 148 at a later date.

add query arg urls also need to be sanitized for redirects too. I've used the old name as `sanitize_url()` was only undeprecated in a recent version of WP.

```suggestion
		$redirect_to = add_query_arg( urlencode( "bulk_{$action}" ), count( $attachment_ids ), $redirect_to );
		return esc_url_raw( $redirect_to );
```